td/tests/tracking/test_views.py

```diff
@@ -0,0 +1,966 @@
+            "is_lead0": True,
+            "speaks_gl0": True,
+            "material0": "Test Material",
+            "licensed0": True,
+        }
+        self.request = RequestFactory().post('/tracking/mc-event/new/', post_data)
+        self.view = setup_view(MultiCharterEventView(), self.request)
+        # For some reason, super's get_context_data() will look for the following attributes
+        self.view.storage = lambda: None
+        setattr(self.view.storage, "extra_data", {})
+        self.view.steps = lambda: None
+        setattr(self.view.steps, "current", "1")
+        self.view.prefix = "multi_charter_event_form"
+        #
+        context = self.view.get_context_data(mock_form_2)
+        #
+        self.assertEqual(context["translators"][0], {"name": "Test Translator"})
+        self.assertEqual(context["facilitators"][0], {"name": "Test Facilitator", "is_lead": True, "speaks_gl": True})
+        self.assertEqual(context["materials"][0], {"name": "Test Material", "licensed": True})
+
+    @patch("td.tracking.views.Language.objects.get")
+    def test_get_form(self, mock_get):
+        """
+        In step 0 without POST, the form returned should be MultiCharterStarter
+        In step 0 with POST, the form returned should MultiCharterEventForm1
+        In step 1, the form returned should be MultiCharterEventForm2
+        """
+        self.assertIsInstance(self.view.get_form(step=0), MultiCharterStarter)
+
+        # instance_dict will be accessed by an internal code
+        self.view.instance_dict = {"1": Event}
+        self.assertIsInstance(self.view.get_form(step=1), MultiCharterEventForm2)
+
+        language = lambda: None
+        setattr(language, "id", "9999")
+        setattr(language, "ln", "Test Language")
+        setattr(language, "lc", "tl")
+        setattr(language, "lr", "Test Region")
+        setattr(language, "gateway_flag", True)
+        mock_get.return_value = language
+        post_data = {"0-language": "9999"}
+        self.request = RequestFactory().post('/tracking/mc-event/new/', post_data)
+        self.view = setup_view(MultiCharterEventView(), self.request)
+        self.assertIsInstance(self.view.get_form(step="0", data=post_data), MultiCharterEventForm1)
+
+    @patch("td.tracking.views.Charter.objects.get")
+    @patch("td.tracking.views.Event.objects.create")
+    @patch("td.tracking.views.get_next_event_number")
+    @patch("td.tracking.views.check_for_new_items")
+    @patch("td.tracking.views.messages.warning")
+    def test_done(self, mock_warning, mock_new_items, mock_next_number, mock_event_create, mock_charter_get):
+        #
+        mock_cleaned_data = {
+            "0-language_0": "9999",
+            "0-language_1": "8888",
+        }
+        self.view.get_all_cleaned_data = Mock(return_value=mock_cleaned_data)
+        #
+        mock_event = Mock()
+        setattr(mock_event, "save", Mock())
+        setattr(mock_event.hardware, "add", Mock())

```

Comment:
Instead of using setattr here, it might look better to use:

```
mock_event.save = Mock()
mock_event.hardware = Mock()
mock_event.hardware.add = Mock()
```